File: src/main/java/com/github/bjoernpetersen/javafxplayback/JavaFxPlayback.java
package com.github.bjoernpetersen.javafxplayback;

import com.github.bjoernpetersen.jmusicbot.playback.AbstractPlayback;
import java.io.File;
import java.io.IOException;
import javafx.scene.media.Media;
import javafx.scene.media.MediaException;
import javafx.scene.media.MediaPlayer;
import javax.annotation.Nonnull;

class JavaFxPlayback extends AbstractPlayback {

  @Nonnull
  private final MediaPlayer player;

  JavaFxPlayback(@Nonnull File file) throws IOException, MediaException {
    this.player = new MediaPlayer(new Media(file.toURI().toURL().toExternalForm()));
    player.setOnEndOfMedia(this::markDone);
  }

  @Override
  public void play() {
    player.play();
  }

  @Override
  public void pause() {
    player.pause();
  }

  @Override
  public void close() throws Exception {
    player.dispose();
    super.close();
  }
}

File: src/main/kotlin/net/bjoernpetersen/javafxplayback/JavaFxPlayback.kt
package net.bjoernpetersen.javafxplayback

import javafx.scene.media.Media
import javafx.scene.media.MediaException
import javafx.scene.media.MediaPlayer
import net.bjoernpetersen.musicbot.spi.plugin.AbstractPlayback
import java.io.File
import java.io.IOException

internal class JavaFxPlayback @Throws(IOException::class, MediaException::class)
constructor(file: File) : AbstractPlayback() {

    private val player = MediaPlayer(Media(file.toURI().toURL().toExternalForm())).also {
        it.setOnEndOfMedia { markDone() }
    }

    override fun play() {
        player.play()
    }

    override fun pause() {
        player.pause()
    }

    @Throws(Exception::class)
    override fun close() {
        player.dispose()
        super.close()
    }
}
